cpp/src/sampling/temporal_sampling_impl.hpp

```diff
@@ -0,0 +1,1108 @@
+            : std::nullopt;
+        auto edge_ids =
+          (edge_id_view) ? std::make_optional(std::move(
+                             std::get<rmm::device_uvector<edge_t>>(sampled_edge_properties[pos++])))
+                         : std::nullopt;
+        auto edge_types =
+          (edge_type_view)
+            ? std::make_optional(std::move(
+                std::get<rmm::device_uvector<edge_type_t>>(sampled_edge_properties[pos++])))
+            : std::nullopt;
+        auto edge_start_times =
+          std::move(std::get<rmm::device_uvector<edge_time_t>>(sampled_edge_properties[pos++]));
+        auto edge_end_times =
+          (edge_end_time_view)
+            ? std::make_optional(std::move(
+                std::get<rmm::device_uvector<edge_time_t>>(sampled_edge_properties[pos++])))
+            : std::nullopt;
+
+        if (weights) { (*result_weight_vectors).push_back(std::move(*weights)); }
+        if (edge_ids) { (*result_edge_id_vectors).push_back(std::move(*edge_ids)); }
+        if (edge_types) { (*result_edge_type_vectors).push_back(std::move(*edge_types)); }
+        (*result_edge_start_time_vectors).push_back(std::move(edge_start_times));
+        if (edge_end_times) {
+          (*result_edge_end_time_vectors).push_back(std::move(*edge_end_times));
+        }
+        if (labels) { (*result_label_vectors).push_back(std::move(*labels)); }
+
+        next_frontier_vertex_spans.push_back(raft::device_span<vertex_t const>{
+          result_dst_vectors.back().data(), result_dst_vectors.back().size()});
+        next_frontier_vertex_time_spans->push_back(
+          raft::device_span<edge_time_t const>{result_edge_start_time_vectors->back().data(),
+                                               result_edge_start_time_vectors->back().size()});
+        if (next_frontier_vertex_label_spans) {
+          next_frontier_vertex_label_spans->push_back(raft::device_span<label_t const>{
+            result_label_vectors->back().data(), result_label_vectors->back().size()});
+        }
+      }
+
+      if (has_duplicates_size > 0) {
+        std::vector<edge_arithmetic_property_view_t<edge_t, vertex_t>> edge_property_views{};
+
+        if (edge_weight_view) edge_property_views.push_back(*edge_weight_view);
+        if (edge_id_view) edge_property_views.push_back(*edge_id_view);
+        if (edge_type_view) edge_property_views.push_back(*edge_type_view);
+        edge_property_views.push_back(edge_start_time_view);
+        if (edge_end_time_view) edge_property_views.push_back(*edge_end_time_view);
+
+        auto [srcs, dsts, sampled_edge_properties, labels] =
+          temporal_sample_edges<vertex_t, edge_t, edge_time_t, multi_gpu>(
+            handle,
+            rng_state,
+            graph_view,
+            raft::host_span<edge_arithmetic_property_view_t<edge_t, vertex_t>>{
+              edge_property_views.data(), edge_property_views.size()},
+            edge_start_time_view,
```

Comment:
What about edge_end_time_view here?

Reply:
Or are we filtering based on edge_end_time later?

Reply:
As mentioned above, edge_end_time is not used for filtering.